### DISCUSSION TITLE
Batching with GraphQL

### DISCUSSION THREAD

**Discussion by UbiquitousBear** [quarkusio/quarkus#39220] [Q&A] [answered]
I am trying to implement batching with GraphQL.

I currently have queries: `allComponents`, `getComponent`, the former returning a collection of items. Each item has a property, `componentType`, which is a separate entity. 

How do I set up batching?

The current code I have is:

```
@Query
        public Uni<Component> getComponent(@Name("id") String id) {
            try {
                return componentService.getComponent(entityReferenceDecoder.getEntityReferenceFromString(id));
            } catch (RuntimeException t) {
                throw new ApiException(t.getMessage());
            }
        }

        public Uni<ComponentType> type(@Source Component component) {
            try {
                return sessionFactory.openSession().chain(() -> componentTypeService.fromEntityReference(entityReferenceDecoder.getEntityReferenceFromString(component.getType().getEntityReference())));
            } catch (RuntimeException t) {
                throw new ApiException(t.getMessage());
            }
        }

@Query("allComponents")
        public Uni<List<Component>> getAllComponents() {
            return componentService.getAllComponents();
        }

public List<List<ComponentType>> componentType(@Source List<Component> components) {
            return List.of(List.of());
        }
```

However, when calling the query:

```
query allComponents {
  allComponents {
    name, 
    type {
      name
    }
  }
}
```

I can't seem to get batching working, I'd expect an empty name value (given I'm using `allComponents`.

[a comment by an automated account is not shown]

**Comment by phillip-kruger** [comment]
You can change the `public Uni<ComponentType> type(@Source Component component)` method  to `public Uni<ListComponentType>> type(@Source List<Component> component)`

see https://www.youtube.com/live/PHWOzzusfrY?feature=shared&t=2752

**Reply by UbiquitousBear** [reply]
@phillip-kruger  - thanks for this: maybe I'm using it incorrectly, I now have a method from teh signature above, however, I'm getting a data exception: `org.hibernate.LazyInitializationException: Unable to perform requested lazy initialization [net.shamil.dxtower.entities.ComponentType.version] - session is closed and settings disallow loading outside the Session`, which makes me believe that the method isn't being used?

This is what my class looks like:

```
@GraphQLApi
@ApplicationScoped
@AllArgsConstructor
@WithTransaction
public class ComponentResource {

    private final ComponentService componentService;
    private final ComponentTypeService componentTypeService;
    private final EntityReferenceDecoder entityReferenceDecoder;
    private final Mutiny.SessionFactory sessionFactory;


    @Query("allComponents")
    public Uni<List<Component>> getAllComponents() {
        return componentService.getAllComponents();
    }

    @Query
    public Uni<Component> getComponent(@Name("id") String id) {
        return ...
    }

    public Uni<List<ComponentType>> type(@Source List<Component> components) {
        return ...
    }
}
```

**Reply by phillip-kruger** [reply]
Can you share how ComponentType looks like ?

**Reply by UbiquitousBear** [reply]
@phillip-kruger 

```
@Entity
@Data
@Cacheable
@Cache(usage = CacheConcurrencyStrategy.READ_WRITE, region = "entities")
@EqualsAndHashCode(callSuper = false)
public class ComponentType extends AbstractEntity {
    @Id
    @Tsid
    @JsonbTransient
    private Long id;
    private String name;
    private String description;
    @Version
    @JsonbTransient
    private long version = 0L;
}
```

**Reply by phillip-kruger** [reply]
Please also share AbstractEntity and Component

**Reply by UbiquitousBear** [reply]
```
@Setter
public abstract class AbstractEntity implements EntityWithReference {

    protected String entityReference;

    public long id() throws RuntimeException {
        Field idField = Stream.of(getClass().getDeclaredFields())
        .filter(field -> field.isAnnotationPresent(Id.class))
        .findFirst()
        .get();

        idField.setAccessible(true);
        try {
            return (long) idField.get(this);
        } catch (IllegalArgumentException | IllegalAccessException e) {
            throw new RuntimeException(e);
        }
    }

    @JsonbTransient
    public String typePrefix() {
        return getClass().getSimpleName().toLowerCase();
    }

    @JsonbProperty
    public String getEntityReference() throws RuntimeException {
        if (Objects.nonNull(entityReference)) {
            return entityReference;
        }

        return EntityReference.builder()
                .partition("global")
                .service("dcc")
                .resourceType(typePrefix())
                .resourceId(String.valueOf(TSID.from(id())).toLowerCase())
                .build()
                .toString();
    }
}
```


and 

```
@NamedEntityGraph(
        name = "fully-hydrated-component",
        attributeNodes = {
                @NamedAttributeNode("type"),
                @NamedAttributeNode("attributes"),
                @NamedAttributeNode("relationships"),
                @NamedAttributeNode("resources"),
                @NamedAttributeNode("scorecards"),
                @NamedAttributeNode("owningTeam"),
                @NamedAttributeNode("lifecycleStage")
        }
)
@Entity
@Data
@Cache(usage = CacheConcurrencyStrategy.READ_WRITE, region = "entities")
@Cacheable
@EqualsAndHashCode(callSuper = false)
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Component extends AbstractEntity {
    @Id
    @Tsid
    @JsonbTransient
    private Long id;
    @ManyToOne(fetch = FetchType.LAZY)
    private ComponentType type;
    @NotBlank
    private String name;
    private String description;
    @Convert(converter = MapJsonConverter.class)
    @Column(columnDefinition = "jsonb")
    @Builder.Default
    private Map<String, String> attributes = new HashMap<>();
    @OneToMany
    private Set<Relationship> relationships;
    @Convert(converter = MapJsonConverter.class)
    @Column(columnDefinition = "jsonb")
    @Builder.Default
    private Map<String, String> resources = new HashMap<>();
    @OneToMany
    @Builder.Default
    private Set<Scorecard> scorecards = new HashSet<>();
    private Integer tierLevel;
    @ManyToOne(fetch = FetchType.LAZY)
    @Nullable
    private Team owningTeam;
    @Nullable
    @Enumerated(EnumType.STRING)
    private LifecycleStage lifecycleStage;
    @Version
    @JsonbTransient
    private long version;

    @Getter
    public enum LifecycleStage {
        PRE_RELEASE,
        IN_PRODUCTION,
        DEPRECATED,
        END_OF_LIFE
    }
}
```

**Reply by phillip-kruger** [reply]
Please try 2 things to see if we can find a workaround:

1) Try without Uni (so blocking)

This should work. If it does not please log an issue.

2) Change the 
```
@ManyToOne(fetch = FetchType.LAZY)
private ComponentType type;
```

to fetch eager

Try with Uni, if this works, please log a bug, as we have an issue with Unis then

/cc @jmartisk

**Reply by UbiquitousBear** [reply]
@phillip-kruger 

1. the service (and repository) return `Uni` - is it safe to try and pull the `List<ComponentType>` out of the `Uni`? How do I do that? I've written a blocking method, but now getting complaints that I'm not in a session - guess I can't use the `Mutiny.SessionFactory`? 

2. Does not work after setting Eager: 
```
org.hibernate.LazyInitializationException: failed to lazily initialize a collection of role: net.shamil.dxtower.entities.Component.relationships: could not initialize proxy - no Session
```

**Reply by phillip-kruger** [reply]
Might be easier to create a small reproducer that myself or @jmartisk can look at. Will that be possible for you?

**Reply by UbiquitousBear** [reply]
@phillip-kruger if it's ok - I'll push up my code to GH and grant you and @jmartisk (and whoever else) access?

**Reply by phillip-kruger** [reply]
Is it runnable? I.e.we would not need other jars from a private repo or external services like db etc ?

**Reply by UbiquitousBear** [reply]
@phillip-kruger `quarkus dev` is the only thing I use, no external deps. It uses a DB, but a dev db that quarkus provides.

**Reply by phillip-kruger** [reply]
Ok great, then that should work

**Reply by robp94** [reply]
The return list should be of the same size as the parameter list. For testing, just map it to some dummy or null. Then it should work. At least for us it does.

**Reply by UbiquitousBear** [reply]
Added you guys @phillip-kruger 

@robp94 I am aware, per [this](https://github.com/quarkusio/quarkus/discussions/29330) discussion; though I am unsure if this is the issue here? Even with 1 item in the parameter, it fails.

**Reply by robp94** [reply]
Does it work without using hibernate in the batch resolver? @phillip-kruger are transactions actually supported by graphql? https://github.com/smallrye/smallrye-graphql/issues/401

**Reply by UbiquitousBear** [reply]
@robp94 how would I not use Hibernate - just return a manual object? Would you like access to the repo?

Surprisingly, it doesn't - I guess for some reason.. I wonder if my `public Uni<List<ComponentType>> type(@Source List<Component> components)` is being called at all?

**Reply by robp94** [reply]
Exactly, just for testing. The old transaction implementation for hibernate reactive was a wrapper around the methods. So the transaction was closed after the query. There is now a new one based on hibernate but i do not know how it works.

Based on the exception it could be this problem. Therefore my question if smallrye graphl supports transactions or not.

**Reply by UbiquitousBear** [reply]
@robp94 I think you're onto it, somehow...

```
public Uni<List<ComponentType>> type(@Source List<Component> components) {
        try {

            return Uni.createFrom().item(components.stream().map(c -> ComponentType.builder().id(entityReferenceDecoder.getEntityIdFromReference(c.getType().getEntityReference())).build()).collect(Collectors.toList()));
//            return sessionFactory.openSession().chain(() -> componentTypeService.byEntityReferences(components.stream().map(c -> entityReferenceDecoder.getEntityReferenceFromString(c.getType().getEntityReference())).toList()));
        } catch (RuntimeException ex) {
            throw new ApiException(ex);
        }
    }
```

the GraphQL query doesn't return the `type` (it's null), however, I see another exception: 

```
org.hibernate.LazyInitializationException: failed to lazily initialize a collection of role: net.shamil.dxtower.entities.Component.relationships: could not initialize proxy - no Session
```

Looks like I'm not handling transactions correctly?

**Reply by UbiquitousBear** [reply]
Taking another look, I don't see why I should be seeing `org.hibernate.LazyInitializationException: Unable to perform requested lazy initialization [net.shamil.dxtower.entities.ComponentType.version] - session is closed and settings disallow loading outside the Session` if `public Uni<List<ComponentType>> type(@Source List<Component> components)` is returning a manually created object; that doesn't make sense to me?

**Reply by robp94** [reply]
I am no expert on hibernate (reactive) maybe @phillip-kruger knows more about it or can ping some one who knows.

**Reply by robp94** [reply]
Did you try to return a list of nulls?

**Reply by UbiquitousBear** [reply]
Yep: 

```
org.hibernate.LazyInitializationException: Unable to perform requested lazy initialization [net.shamil.dxtower.entities.ComponentType.name] - session is closed and settings disallow loading outside the Session

public Uni<List<ComponentType>> type(@Source List<Component> components) {
        try {
            List<ComponentType> list = new LinkedList<>();
            components.forEach(c -> list.add(null));
            return Uni.createFrom().item(list);
        } catch (RuntimeException ex) {
            throw new ApiException(ex);
        }
    }
```

**Reply by robp94** [reply]
Ok may last idea would be the transactional annotation. Maybe only annotate query but not the class and not the method with the `@source` annotation.

**Reply by UbiquitousBear** [reply]
I've added the `@WithTransaction` to the Service classes, still nope: 

```
org.hibernate.LazyInitializationException: Unable to perform requested lazy initialization [net.shamil.dxtower.entities.ComponentType.name] - session is closed and settings disallow loading outside the Session
```

I'll push this change up to the repo - do you want to have access to the repo?

**Reply by phillip-kruger** [reply]
I am running your example, however I need test data. Can you share how you would populate the DB to test the query above ?

**Reply by UbiquitousBear** [reply]
@phillip-kruger oops, totally:

Step 1: Create the `Component Entity` at `/graphql`:

```
{
	"query": "mutation newComponent($input:CreateComponentMutationInput){newComponent(component:$input){entityReference,name,description}}",
	"variables": {
		"input": {
			"name": "A Service",
			"description": "A new service",
			"tierLevel": "1",
			"typeReference": "snw:global:dcc:componenttype:0em1pt8w7qt11"
		}
	}
}

```

Step 2: Pull out all components:

```
query allComponents {
  allComponents {
    name,
    entityReference,
    type {
      name
    }
  }
}
```

This query has `type:name` which should be pulled.

**Reply by phillip-kruger** [reply]
Thanks :)

**Reply by phillip-kruger** [reply]
@UbiquitousBear this example has got too many things going to narrow this down. Can you create the simplest reproducer you can. I don't know if the issue is somewhere in the JPAStreamer code or Panache. If I hardcode a list in the @Source method, it works fine. If I add a getAllComponentTypes query, that breaks too. Seems like there is an issue with getting the Component types

**Reply by UbiquitousBear** [reply]
@phillip-kruger We can skip the JPAStreamer code and use Panache directly (`list()`): I've [pushed up code](https://github.com/UbiquitousBear/dx-tower/commit/5742a9a07c813fcd72f3e188de6c484272a36993) the change this.  A quick retest shows the same issue.

**Reply by phillip-kruger** [reply]
Ok I'll have a look again in the morning

**Reply by UbiquitousBear** [reply]
Thanks @phillip-kruger, I've made a small adjustment to the queries so we always return what has been asked for, new code has been pushed up; have a good evening!

**Reply by phillip-kruger** [reply, marked as answer]
OK. There is a few things going on here.

1) Lombok. 

The original error you are getting is due to Lombok's equals and hashcode. I removed the `@EqualsAndHashCode(callSuper = false)` and `@Data` on both `Component` and `ComponentType` and added `@Getter @Setter`

2) Wrong parameter. 

The parameter passes into the GraphQL `@Source` method needs to match the response type of the `@Query`. So I changed 

`public Uni<List<ComponentType>> batchedType(@Source List<Component> components)`
to
`public Uni<List<ComponentType>> batchedType(@Source Uni<List<Component>> components)`

in `ComponentResource`

3) Conflicting field names.

So you have:

```
@ManyToOne(fetch = FetchType.LAZY)
private ComponentType type;
```

on `Component`

So in GraphQL there is a `type` field on component.

Then you also add a `type` field via the `@Source` annotation, so now there are two fields with the same name. 

(@jmartisk we need to fail or at least warn when this happens - I created an issue for this: https://github.com/smallrye/smallrye-graphql/issues/2064)

Now you have to choose one or the other:

A) Using `@ManyToOne`
If the field are on the Component as a ManyToOne, marking it as `EAGER` works. No GraphQL Batching is not involved. When not using Uni (so non-blocking) and the non blocking db driver, LAZY should work. I created another issues to look into getting LAZY to work with non-blocking driver (if possible) https://github.com/smallrye/smallrye-graphql/issues/2063

B) Using GraphQL Batching
In this case there should not be a type field on the `Component` Object, only a `Source` that will fetch it. I am not sure if this will work as the current example are using the type field from the `Component`. 

If you do change the implementation to use B, please let me know if that worked or if we need another issue

Hope this helps.

**Reply by UbiquitousBear** [reply]
@phillip-kruger  sounds good - I think in this case, it makes more sense to set the fetch type to eager, and omit the type function in the resource - thanks :)
